packages/promise-tracker/lib/index.ts

```diff
@@ -0,0 +1,104 @@
+const _outstandingPromiseInstanceMap = new WeakMap<Object, Set<Promise<any>>>();
+const _outstandingPromiseMap = new Map<Promise<any>, boolean>();
+
+export function tracked<T extends Object>(
+  target: T,
+  key: string | Symbol,
+  descriptor: PropertyDescriptor
+) {
+  if (typeof descriptor.value === "function") {
+    const impl = descriptor.value;
+
+    // wrapper function that keeps a list of outstanding promises for each
+    // instance of DashboardMessageBusClient
+    function newImpl(...args: any[]) {
+      const returnValue = impl.apply(this, args);
+
+      // only track async functions
+      if (returnValue.then && typeof returnValue.then === "function") {
+        const trackedTask = returnValue
+          .then((result: any) => {
+            _cleanUpTrackedTask(this, trackedTask);
+            return result;
+          })
+          .catch((err: any) => {
+            _cleanUpTrackedTask(this, trackedTask);
+            throw err;
+          });
+
+        _trackTask(this, trackedTask);
+        return trackedTask;
+      }
+
+      return returnValue;
+    }
+
+    return {
+      ...descriptor,
+      value: newImpl
+    };
+  }
+
+  return descriptor;
+}
+
+export interface WaitForOutstandingPromiseOptions<T> {
+  catchHandler?: (err?: any) => void;
+  target?: T;
+}
+
+export async function waitForOutstandingPromises<T extends Object>(
+  options?: WaitForOutstandingPromiseOptions<T>
+) {
+  const { target } = options || {};
+  let { catchHandler } = options || {};
+
+  const iterable =
+    (target
+      ? _outstandingPromiseInstanceMap.get(target)?.values()
+      : _outstandingPromiseMap.keys()) || [];
+
+  catchHandler = catchHandler || (() => {});
```

Comment:
Should the default be to swallow errors? Perhaps we should use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AggregateError and throw at the end?

Reply:
Yes, per above comment. The goal is to allow any "best effort" tasks to complete, even if they complete unsuccessfully (e.g. sending events to the dashboard message bus). The message bus might not be running, however. That would cause the client to throw, but nothing needs to be done with that error, as it's not considered exceptional.